### ISSUE TITLE
Allow interceptions and bandwidth throttling of TestServer

### ISSUE THREAD

**Issue by weitzhandler** [aspnet/Hosting#1293]
Hi,

I'm using `Microsoft.AspNetCore.TestHost.TestServer` to test my ASP.NET Core Web API (my real client is WPF).

I tried setting the URL of both the server and the test client in my test fixture as follows (using [DryIoc](https://bitbucket.org/dadhi/dryioc)):

```c#
public class ContainerFixture : IDisposable
{
  public ContainerFixture()
  {
    Container = new Container();
    RegisterTypes(Container);
  }

  private void RegisterTypes(IContainer container)
  {
    container.RegisterDelegate(r =>
      new TestServer(new WebHostBuilder()
      .UseStartup<Startup>()
      .UseUrls(Constants.Url)),
      Reuse.Singleton);

    container.RegisterDelegate(r =>
    {
      var client = r
      .Resolve<TestServer>()
      .CreateClient();

      client.BaseAddress = new Uri(Constants.Url);

      return client;
    },
    Reuse.Singleton);
  }

  public IContainer Container { get; }
  public void Dispose() => Container.Dispose();
}
```

But nothing appears in Fiddler.

I'd hereby like to suggest you to implement the following features:

- Make it easier to interfere and intercept transactions between the test client and test server
- Allow throttling the bandwidth as to simulate a slow connection between client and server

**Comment by Tratcher** [comment]
UseUrls and Fiddler do not apply to TestServer because it never creates a socket, it's all processed in memory.

You can add an interception layer by calling `CreateHandler()` and wrapping it in your own HttpMessageHandler with your throttling logic.

**Comment by weitzhandler** [comment]
Thanks for your quick response!
I'm testing a file upload-download server, and the best way to test the performance is by throttling the speed.

- Any ideas on how to throttle bandwidth in a message handler?
- So once I have a `DelegatingHandler` or so that tunnels down to the `TestServer` provided handler, can I then create my own `HttpClient` (instead of `CreateClient`)?

**Comment by Tratcher** [comment]
In your handler you wrap and replace the HttpRequestMessage.Content and override CreateContentReadStream to add a stream wrapper that with the built in delay you want. You can do something similar for the response.

And yes you then create your own HttpClient over your custom handler, that's all CreateClient does.

**Comment by muratg** [comment]
We are closing this issue because no further action is planned for this issue. If you still have any issues or questions, please log a new issue with any additional details that you have.

**Comment by weitzhandler** [comment]
BTW, the `TestServer` isn't the right idea to test file uploads.

The controller streams will be completely different when running in concrete environment.
This should be either addressed or clearly pointed out in the front page.
I just finished a whole file-upload API using the `TestServer` only to realize all the stream types and overall behavior I relied on is to be totally different in a production or other server.

**Comment by weitzhandler** [comment]
@muratg 

BTW, the `TestServer` isn't the right idea to test file uploads.

The controller streams will be completely different when running in concrete environment.
This should be either addressed or clearly pointed out in the front page.
I just finished a whole file-upload API using the `TestServer` only to realize all the stream types and overall behavior I relied on is to be totally different in a production or other server.
